src/locale/localed.c

```diff
@@ -395,6 +399,37 @@ static int method_set_locale(sd_bus_message *m, void *userdata, sd_bus_error *er
         for (LocaleVariable p = 0; p < _VARIABLE_LC_MAX; p++)
                 free_and_replace(c->locale[p], new_locale[p]);
 
+        /* Generate locale in case it is missing and the system is using locale-gen */
+        if (use_localegen) {
+                for (LocaleVariable p = 0; p < _VARIABLE_LC_MAX; p++) {
+                        if (isempty(c->locale[p]))
+                                continue;
+                        if (locale_is_installed(c->locale[p]))
+                                continue;
+
+                        r = locale_gen_enable_locale(c->locale[p]);
+                        if (r < 0) {
+                                if (r != -EINVAL) {
+                                        log_error_errno(r, "Failed to enable locale generation: %m");
+                                        return sd_bus_error_set_errnof(error, r, "Failed to enable locale generation: %m");
+                                }
+                        } else {
+                                r = locale_gen_run();
+                                if (r != 0) {
+                                        log_error_errno(r, "Failed to generate locale: %m");
+                                        return sd_bus_error_set_errnof(error, r, "Failed to generate locale: %m");
+                                }
+                        }
+
+                        /* Sanity check to ensure the new locale is installed and usable now */
+                        if (locale_is_installed(c->locale[p]) <= 0) {
+                                log_warning("Failed to generate a working locale with locale-gen for %s", c->locale[p]);
+                                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Specified locale is not installed: %s", c->locale[p]);
+                        }
+                }
+        }
```

Comment:
given how complex this is, maybe split out this loop into a helper function on its own?

Reply:
It has grown quite a bit for what I initially though would be a much, much smaller change indeed!